osquery/events/windows/evtsubscription.cpp

```diff
@@ -0,0 +1,163 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+
+#include <osquery/events/windows/evtsubscription.h>
+#include <osquery/logger.h>
+#include <osquery/utils/conversions/windows/strings.h>
+
+namespace osquery {
+// Note: Windows ignores the exit code of this function
+DWORD WINAPI EvtSubscriptionCallbackDispatcher(
+    EVT_SUBSCRIBE_NOTIFY_ACTION action, PVOID context, EVT_HANDLE event) {
+  auto& subscription = *static_cast<EvtSubscription*>(context);
+
+  if (action == EvtSubscribeActionError) {
+    LOG(ERROR)
+        << "Windows event callback error 'EvtSubscribeActionError' for channel "
+        << subscription.channel();
+
+    return 0;
+
+  } else if (action != EvtSubscribeActionDeliver) {
+    LOG(ERROR) << "Windows event callback invoked with invalid action value "
+                  "for channel "
+               << subscription.channel();
+
+    return 0;
+  }
+
+  subscription.processEvent(event);
+  return 0U;
+}
+
+struct EvtSubscription::PrivateData final {
+  EVT_HANDLE handle{nullptr};
+  std::string channel;
+
+  EventList event_list;
+  std::mutex event_list_mutex;
+  std::condition_variable event_list_cv;
+};
+
+Status EvtSubscription::create(EvtSubscription::Ref& obj,
+                               const std::string& channel) {
+  obj.reset();
+
+  try {
+    obj.reset(new EvtSubscription(channel));
+    return Status::success();
+
+  } catch (const std::bad_alloc&) {
+    return Status::failure("Memory allocation failure");
+
+  } catch (const Status& status) {
+    return status;
+  }
+}
+
+EvtSubscription::~EvtSubscription() {
+  EvtClose(d->handle);
+}
+
+const std::string EvtSubscription::channel() const {
+  return d->channel;
+}
+
+EvtSubscription::EventList EvtSubscription::getEvents() {
+  EventList event_list;
+
+  {
+    std::unique_lock<std::mutex> lock(d->event_list_mutex);
+
+    auto ready = d->event_list_cv.wait_for(
+        lock, std::chrono::seconds(1U), [=]() -> bool {
```

Comment:
In a lot of other parts of osquery we block for 200 milliseconds. I think we should make these waiting decisions based on what makes sense for each use case but I wanted to mention the "common timeout".